### PR TITLE
Mykola Melnyk Java Lecture 8 

### PR DESCRIPTION
**Pull request by Mykola-Melnyk**

### CODE REVIEW

**Comment by tromanyuk** [comment]
Mykola, please resolve merge conflicts (I think you can do the same as with the previous PR).

**Comment by Mykola-Melnyk** [comment]
> Mykola, please resolve merge conflicts (I think you can do the same as with the previous PR).

Dear Taras,
I've resolved the conflicts.
Sorry for the inconveniences.

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_1.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.zoolatech.lecture7.tasks._2;
+
+public class Calc_v_1 implements Calc {
+    private volatile int storedValue;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_1.java:4]: you don't need to use `volatile` if the access to a variable is protected with `synchronized`

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package com.zoolatech.lecture7.tasks._2;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class Calc_v_2 implements Calc {
+    private volatile int storedValue;
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java:7]: same as in a previous version:
1. volatile can be removed
2. getter should also use the lock object

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Test_Task2.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.zoolatech.lecture7.tasks._2;
+
+public class Test_Task2 {
+    public static void main(String[] args) throws InterruptedException {
+        Calc_v_3 calc3 = new Calc_v_3(1);
+        Calc_v_2 calc2 = new Calc_v_2(1);
+        Calc_v_1 calc1 = new Calc_v_1(1);
+        Test_Task2.testCalc(calc3);
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Test_Task2.java:8]: you don't need to specify the class name to access a static method in the same class

File: test_demo3.json [hunk 4]
```diff
@@ -0,0 +1 @@
+test text2
```
Comment (tromanyuk) [inline test_demo3.json:1]: please remove files test_demo3.json and test_obj3.json from the pull request

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_1.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    public Calc_v_1(int value) {
+        storedValue = value;
+    }
+    public int getStoredValue() {
```
[the remark this answers is not in the document]
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_1.java:8]: Done.

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java (hunk 2 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java:7]: Done

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Test_Task2.java (hunk 3 above)
Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Test_Task2.java:8]: Removed redundant class names, thank you

File: java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class Calc_v_2 implements Calc {
+    private int storedValue;
+    private Lock calcLock = new ReentrantLock();
+    public Calc_v_2(int value) {
+        storedValue = value;
+    }
+    public synchronized int getStoredValue() {
```
Comment (tromanyuk) [inline java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java:12]: please replace `synchronized` with a lock

Comment (Mykola-Melnyk) [reply java/mykola-melnyk/src/com/zoolatech/lecture7/tasks/_2/Calc_v_2.java]: Sorry, forgot that this is done with the lock. Done.

**Review by tromanyuk — APPROVED** [review]

**Pull request merged**